Docs: TUF docs Analysis and Improvement (#261)
cncf/techdocs: analyses/0012-TUF/analysis.md

---
title: TUF Documentation Analysis
tags: TUF
created: 2024-06-17
modified: YYYY-MM-DD
author: Sandra Dindi (@Dindihub)
---

<!-- markdownlint-disable no-duplicate-heading no-empty-links -->

## Introduction

This document analyzes the effectiveness and completeness of
[The Update Framework](https://theupdateframework.io) (TUF) open source software
(OSS) project's documentation and website. It is funded by the CNCF Foundation
as part of its overall effort to incubate, grow, and graduate open source cloud
native software projects.

According to CNCF best practices guidelines, effective documentation is a
prerequisite for program graduation. The documentation analysis is the first
step of a CNCF process aimed at assisting projects with their documentation
efforts.

### Purpose

This document analyzes the current state of **The Update Framework (TUF)**
documentation. It provides project leaders with an informed understanding of
potential problems in current project documentation. A second
[implementation](./implementation.md) document, outlines an actionable plan for

> improvement. A third document, the [issues](./issues.md) outlines issues to be

Reviewer: ```suggestion
improvement. A third [issues](./issues.md) document, outlines the issues to be
```

This analysis measures documentation against CNCF project maturity standards,
and suggests possible improvements. In most cases there is more than one way to
do things. Few recommendations here are meant to be prescriptive. Rather, the
recommended implementations represent the reviewers' experience with how to
apply documentation best practices. In other words, borrowing terminology from
the lexicon of [RFCs][rfc-spec], the changes described here should be understood
as "recommended" or "should" at the strongest, and "optional" or "may" in many
cases. Any "must" or "required" actions are clearly denoted as such, and pertain
to legal requirements such as copyright and licensing issues.

## Project documentation

| Criteria                   | Rating (1-5) |
| -------------------------- | ------------ |
| Information architecture   | 2            |
| New user content           | 1            |
| Content maintainability    | 3            |
| Content creation processes | 1            |
| Inclusive language         | 3            |

Scale:

1 = Not present

2 = Needs improvement

3 = Meets standards

4 = Meets or exceeds standards

5 = Exemplary

### Summary of issues

> - Information is repeated is several pages of the website. For example, the TUF

Reviewer: ```suggestion
- Information is repeated on several pages of the website. For example, the TUF
```
I use "on" instead of "in" because, "in" is best suitable when referencing a location of something within another thing. Though they are both prepositions, it's very easy to mix them.

>   Specification file is referenced in more than one page. Some pages can be

Reviewer: ```suggestion
  Specification file is referenced on more than one page. Some pages can be
```

  related to their use case. For example, information in the sections should
  correspond with the label of the section i.e _About_ should only contain
  introductory content about the project.

- The information available for news users and contributors is not satisfactory.
  There are no step-to-step guides and tutorials for new users to get started
  with the software, neither are there contributor guidelines.

#### Information architecture

- There is an _overview_ section explaining what TUF is and its use cases. The
  features, the metadata and roles are well explained with examples for each
  metadata.

- Repetition of content on different pages makes content confusing

- Re-organise content to make it easier to follow

- There are no tutorials for specific feature implementation. But, there are
  videos explaining various use cases.

- There aren't specific task-based guides for features. The available guide for
  implementing the specification is part of a larger document labelled
  [Specification(latest)](https://theupdateframework.github.io/specification/latest)

- There is a FAQ and reporting issues sections for troubleshooting.

> - There is a a well detailed API reference for multiple TUF APIs

Reviewer: Great work so far, @Dindihub. Here are a few suggestions.

```suggestion
- There is a well detailed API reference for multiple TUF APIs
```

- README on
  [theupdateframework.io](https://github.com/theupdateframework/theupdateframework.io)

>   repo is empty with little information about content of the repo.

Reviewer: ```suggestion
  repo is empty with little information about the content of the repo.
```

#### New user content

- There is a _Getting started_ section on the website with information about
  main features. But there's repetition of content from other pages. For example
  the _Specification (latest)_ and _index_ include information present in the
  project and overview sections. This section is confusing for new users.

- The documentation repo does not have a contributor guide for new users to get
  started

- There isin't any documentation labeled 'Installation guide'. Instead,
  installation instructions are part of a larger document labelled

>   [The Update Framework specification](https://theupdateframework.github.io/specification/latest)

Author: I saw this section on the spec  https://theupdateframework.github.io/specification/latest/#detailed-client-workflow and confused it for some sort of Installation/configuration guide. Thanks for clearing that up

- TUF docs do not provide information about application-specific/OS
  functionality in system updates. Instead the documentation states that TUF
  provides a secure way for applications to obtain and verify files being
  distributed by trusted parties.

- There is sample code in the content that can be easily copy-pasted on other
  platforms.

- It's not clear whether the documentation is up-to-date as there are no dates
  on docs or release notes published.The only document on the site with dates is
  the
  [Specification file](https://theupdateframework.github.io/specification/latest)

#### Content maintainability & site mechanics

- The documentation is not searchable. You have to go through the site to find

>   what you are looking for. The only source of naviagation is the menu bar.

Reviewer: ```suggestion
  what you are looking for. The only source of navigation is the menu bar.
```

> - The Docs are managed through docs-as-code site Hugo and content written in

Reviewer: ```suggestion
- The Docs are managed through the docs-as-code site Hugo and content is written in
```

>   markdown. However,it appears the updates are made manually.

Reviewer: ```suggestion
  markdown. However, it appears the updates were made manually.
```

> 

Reviewer 2: @Dindihub ☝️ 

I see a bunch of other comments by @lukpueh that don't seem to have been read and/or processed by you. For each comment like this one by @lukpueh, please mark as resolved or give it a thumbs up when you are done processing it so that we'll know that you've seen it.

(I know that some comments are tagged by GH as "Outdated", but that doesn't confirm to me that you've addressed the comment -- because they can become "Outdated" by mistake.)

#### Content creation processes

- Documentation lacks contribution process guides and information on how to get
  started.

- Documentation lacks procedures for duplicating the documentation locally.

- It's not clear whether the code release process is synced with the

>   documentation creation and updates.

Reviewer 2: @Dindihub - add the subprojects listed above to the analysis doc so that we have the reference (and even if they are out of scope).

- It's not clear who reviews and approves documentation pull requests and
  updates either on the website or repo.

- Information about TUF project maintainers is available on the website but not
  on the project documentation repo.

#### Inclusive language

> - I noted one instance of use of non-recommended words as documented by the

Reviewer: ```suggestion
- I noted one instance of the use of non-recommended words as documented by the
```

  [Inclusive Naming Initiative](https://inclusivenaming.org) website. The word
  _Aborted_ is used in the
  [Specification index tutorial](https://theupdateframework.github.io/specification/latest/#fix-time)

> - There is no use of abliest language like simple,easy in the documentation.

Reviewer: ```suggestion
- There is no use of abliest language like simple, or easy in the documentation.
```

| Criteria                                  | [Rating (1-5)] |
| ----------------------------------------- | -------------- |
| Communication methods documented          | 3              |
| Beginner friendly issue backlog           | 1              |
| “New contributor” getting started content | 1              |
| Project governance documentation          | 3              |

Scale:

1 = Not present

2 = Needs improvement

3 = Meets standards

4 = Meets or exceeds standards

5 = Exemplary

### Summary of Issues

- The documentation does not contain information tailored to contributors.
- Information about TUF communication channels is not visible in the docs repo.
  But the information is available on the website.
- The documentation does not contain information about other project repos and
  their links. Making it harder for contributors to find them.
- The documentation repo issues do not appear to be maintained. There are old
  issues that are still open which may confuse contributors looking for issues
  to work on.
- Issues do not have labels making it hard for contributors to identify suitable
  areas of interest

#### Communication methods documented

> - Information about the TUF **slack channel** is available on the website on

Reviewer: ```suggestion
- Information about the TUF **slack channel** is available on the website in
```

  both the Community and Contact sections. However,this information should be
  visible on the doc repo README for easy access.

- The repo link on the website does not point to the documentation but rather
  the

>   [Python reference implementation](https://github.com/theupdateframework/python-tuf?tab=readme-ov-file)

Reviewer 2: The GitHub link should point to the website repo.
We can link to the community repo from the Community page.

Another possibility is to have two GitHub links, as is done on https://etcd.io (navigate to the footer at the bottom of the page):

- One link is to the organization. From there, a user can find any org-related repo
- The other link is to the website repo

- Information about the mailing list is included in the documentation. There's

>   no information about project meetings. Instead the users are directed to join

Reviewer: ```suggestion
  no information about project meetings. Instead, the users are directed to join
```

  the #TUF channel on CNCF slack.

#### Beginner friendly issue backlog

- Issues on the docs repo are missing labels, making it hard for contributors to
  identify 'Docs' issues.
- New contributors will have a hard time getting started as none of the issues
  are marked "good first issue” label
- Most issues on the docs repo have a title and a detailed description.
- Issues are not maintained for staleness. There are issues opened in 2021 that
  are still open with the information on their status.

#### New contributor getting started content

- There's a community section on the website with information on how to join the

>   community.

Author: > I just checked and it looks like there isn't, or do you mean the link to the slack channel?

I suggested having more communication information in the implementation guide.

- No specific documentation tailored for new users to get started the software

>   on the website or the docs repo.

Author: > Do you mean the software to build the website (Hugo etc.), or the software to "use TUF".

Build the website. I'll make it clearer. Thanks

- The docs provide several channels for reporting issues, contacting the
  maintainers, and slack community.

#### Project governance documentation

- The project governance is well documented on the website. There's sufficient
  informtaion about it's history including leaders,maintainers and contributors.
  However, this information is not visisble in the docs repo README.

### Contributor recommendations

#### Communication methods documented

- The documentation repo link may be included on the site as an entry point for
  the project. The repo can then include information about other repos.
- Include communication channels on the Docs repo README for visibility.

> - Provide information about project meetings e.g a meeting link and calendar.

Reviewer: ```suggestion
- Provide information about project meetings, e.g., a meeting link and calendar.
```

#### Beginner friendly issue backlog

> - Maintain issues, track,close and stale old issues to reduce backlog.

Reviewer: ```suggestion
- Maintain issues, track, close, and stale old issues to reduce backlog.
```

- Label all issues to assist new users and contributors identify areas of
  interest.

#### New contributor getting started content

> - Include instructions on how to contribute on a CONTRIBUTING guideline e.g how

Reviewer: ```suggestion
- Include instructions on how to contribute to a CONTRIBUTING guideline, e.g., how
```

>   to identify issues,forking,cloning, and submitting PRs.

Reviewer: ```suggestion
  to identify issues, forking, cloning, and submitting PRs.
```

| Single-source for all files                 | 4              |
| Meets min website req. (for maturity level) | 3              |
| Usability, accessibility, and design        | 3              |
| Branding and design                         | 4              |
| Case studies/social proof                   | 5              |
| SEO, Analytics, and site-local search       | 2              |
| Maintenance planning                        | 2              |
| A11y plan & implementation                  | 3              |
| Mobile-first plan & impl.                   | 5              |
| HTTPS access & HTTP redirect                | 5              |
| Google Analytics 4 for production only      | 1              |
| Indexing allowed for production server only | 1              |
| Intra-site / local search                   | 1              |
| Account custodians are documented           | 3              |

Scale:

1 = Not present

2 = Needs improvement

3 = Meets standards

4 = Meets or exceeds standards

5 = Exemplary

### Summary of Issues

- The website repo is named _theupdateframework.io_ which makes it harder to
  identify it. Name it according to it's content.

> - There's no guideline or tutorial to assist users to generate the website from

Reviewer: ```suggestion
- There's no guideline or tutorial to assist users in generating the website from
```

  the website repo.
- The available [TUF blog](https://theupdateframework.github.io/python-tuf) page
  is not available on the website. It's hosted on GitHub.
- Information about the website build, tools and how to generate it are not
  available on the website or the docs repo.
- Intra-site search mechanism is not available from the website. The only

>   naviagation option is a menu bar. This makes it difficult to find information.

Reviewer: ```suggestion
  navigation option is a menu bar. This makes it difficult to find information.
```

> - Alot of empty space between the Hero section and the Navbar. Due to this

Reviewer: ```suggestion
- A lot of empty space between the Hero section and the Navbar. Due to this
```

  spacing, information is pushed out of eyelevel. You have to scroll down to
  find it.

#### Single-source requirement

- All source files are in the website repo named
  [theupdateframework.io](https://github.com/theupdateframework/theupdateframework.io)

#### Minimal website requirements

> The website docs analysis is in progess.

Reviewer: ```suggestion
The website docs analysis is in progress.
```

#### Usability, accessibility and devices

- The website follows a mobile-first design with all pages, menu's and features
  visible on smaller screens.
- Website pages are readable
- The website does not provide a text-to-speech option for users.

#### Branding and design

> - There's a recognizable brand for the project , a logo and blue-white color

Reviewer: ```suggestion
- There's a recognizable brand for the project, a logo, and a blue-white color
```

>   scheme used througout the website.

Reviewer: ```suggestion
  scheme used throughout the website.
```

- The website design is good and suitable for reading. However, consider

>   reducing space between the hero section and the Navbar.

Reviewer: ```suggestion
  reducing the space between the hero section and the Navbar.
```

#### Case studies/social proof

- There are case studies documented on the website under the _Adoptions_ page.
- There's a logo, wall of users and participating organizations documented in
  the _Adoptions_ page.

> - No avaliable community talks on the website. They have however provided links

Reviewer: ```suggestion
- No available community talks on the website. They have however provided links
```

  to the community channel.
- The available [TUF blog](https://theupdateframework.github.io/python-tuf) page
  is not available on the website. It's hosted on GitHub

#### SEO, Analytics and site-local search

- It's not clear what analytics are used on the site
- Intra-site search is not available from the website
- There's no mention of custodians of analytics and search.

#### Maintenance planning

- The website runs on Hugo and supported by the community.
- It's not clear who are the maintainers of the website.

#### Other

- The website is accessible via HTTPS

### Website recommendations

#### Single-source requirement

- Rename the website repo _theupdateframework.io_ to a name that reflect it's

>   content e.g _TUFwebsite.io_. It makes it easier to find.

Reviewer: ```suggestion
  content, e.g., _TUFwebsite.io_. It makes it easier to find.
```

#### Branding and design

> - Reduce empty space between hero section and Navbar to bring information to

Reviewer: ```suggestion
- Reduce the empty space between the hero section and Navbar to bring information to
```

  eyelevel. At the moment the information is too far down, you have to scroll to
  find it.

#### SEO, Analytics and site-local search

- Provide intra-site search options such as a search button to make it easier
  for users to search and find information.

#### Case studies/social proof

- Add the [TUF blog](https://theupdateframework.github.io/python-tuf) to the
  website.

#### Maintenance planning

- Identify website maintainers on the site and their roles so users know who to
  contact in case issues arise.

> - Provide information about the website build and how to generate on the docs

Reviewer: ```suggestion
- Provide information about the website build and how to generate it on the docs
```

**[issues](./issues.md) list**.

Readers interested in the current state of the documentation and the reasoning
behind the recommendations should read the section of this document pertaining
to their area of concern:

- [Project documentation]()
- [Contributor documentation]()
- [Website and documentation infrastructure]()

#### Recommendations, requirements, and best practices

This analysis measures documentation against CNCF project maturity standards,
and suggests possible improvements. In most cases there is more than one way to
do things. Few recommendations here are meant to be prescriptive. Rather, the
recommended implementations represent the reviewers' experience with how to
apply documentation best practices. In other words, borrowing terminology from
the lexicon of [RFCs][rfc-spec], the changes described here should be understood
as "recommended" or "should" at the strongest, and "optional" or "may" in many
cases. Any "must" or "required" actions are clearly denoted as such, and pertain
to legal requirements such as copyright and licensing issues.

## Project Documentation

| Criteria                   | Rating (1-5) |
| -------------------------- | ------------ |

> | Information architecture   | 3            |

Author: I agree. After seeing your recommendation to have a separate Docs section. Needs a complete overhaul.

author: Sandra Dindi (@Dindihub)
---

> <!-- markdownlint-disable no-duplicate-heading no-empty-links -->

Reviewer 2: @Dindihub - this line should not be removed. Without it the markdown check fails, see https://github.com/cncf/techdocs/actions/runs/9741513776/job/26895744184?pr=261
